pkg/approver/allowed/evaluator.go

```diff
@@ -0,0 +1,206 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package allowed
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	utilpki "github.com/jetstack/cert-manager/pkg/util/pki"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	policyapi "github.com/cert-manager/policy-approver/pkg/apis/policy/v1alpha1"
+	"github.com/cert-manager/policy-approver/pkg/approver"
+	"github.com/cert-manager/policy-approver/pkg/approver/internal"
+)
+
+// Evaluate evaluates whether the given CertificateRequest conforms to the
+// allowed attributes defined in the policy. The request _must_ conform to
+// _all_ allowed attributes in the policy to be permitted by the passed policy.
+// If the request is denied by the allowed attributes an explanation is
+// returned.
+// An error signals that the policy couldn't be evaluated to completion.
+func (a Allowed) Evaluate(_ context.Context, policy *policyapi.CertificateRequestPolicy, request *cmapi.CertificateRequest) (approver.EvaluationResponse, error) {
+	var (
+		// el will contain a list of policy violations for fields, if there are
+		// items in the list, then the request does not meet the allowed
+		// attributes.
+		el      field.ErrorList
+		allowed = policy.Spec.Allowed
+		fldPath = field.NewPath("spec", "allowed")
+	)
+
+	if allowed == nil {
+		allowed = new(policyapi.CertificateRequestPolicyAllowed)
+	}
+
+	csr, err := utilpki.DecodeX509CertificateRequestBytes(request.Spec.Request)
+	if err != nil {
+		return approver.EvaluationResponse{}, err
+	}
+
+	if len(csr.Subject.CommonName) > 0 {
+		if allowed.CommonName == nil {
+			el = append(el, field.Invalid(fldPath.Child("commonName"), csr.Subject.CommonName, "nil"))
+		} else if !internal.WildcardMatchs(*allowed.CommonName, csr.Subject.CommonName) {
+			el = append(el, field.Invalid(fldPath.Child("commonName"), csr.Subject.CommonName, *allowed.CommonName))
+		}
+	}
+
+	if len(csr.DNSNames) > 0 {
+		if allowed.DNSNames == nil {
+			el = append(el, field.Invalid(fldPath.Child("dnsNames"), csr.DNSNames, "nil"))
+		} else if !internal.WildcardSubset(*allowed.DNSNames, csr.DNSNames) {
+			el = append(el, field.Invalid(fldPath.Child("dnsNames"), csr.DNSNames, strings.Join(*allowed.DNSNames, ", ")))
+		}
+	}
+
+	if len(csr.IPAddresses) > 0 {
+		var ips []string
+		for _, ip := range csr.IPAddresses {
+			ips = append(ips, ip.String())
+		}
+		if allowed.IPAddresses == nil {
+			el = append(el, field.Invalid(fldPath.Child("ipAddresses"), ips, "nil"))
+		} else if !internal.WildcardSubset(*allowed.IPAddresses, ips) {
```

Comment:
Yes, it would be better if you could define subnets in this way. This would likely be another allowed field I think (`ipSubnets`), but this can be done in a followup PR.

Reply:
I imagine the range field could just replace this one (you could specify a single IP with `/32` as for various firewall configs etc), so we don't have two fields for the same thing.